storage/replica_command.go

```diff
@@ -725,11 +769,11 @@ func (r *Replica) HeartbeatTxn(batch engine.Engine, ms *engine.MVCCStats, h roac
 	if ok, err := engine.MVCCGetProto(batch, key, roachpb.ZeroTimestamp, true, nil, &txn); err != nil {
 		return reply, err
 	} else if !ok {
-		// If no existing transaction record was found, initialize to a
-		// shallow copy of the transaction in the request header. We copy
-		// to avoid mutating the original below.
-		// TODO(tschottdorf): double-check that we don't want a clone.
-		txn = *h.Txn
+		// If no existing transaction record was found, skip heartbeat.
+		// This could mean the heartbeat is a delayed relic or it could
+		// mean that the BeginTransaction call was delayed. In either
+		// case, there's no reason to persist a new transaction record.
+		return reply, nil
```

Comment:
ah, I saw that you already changed those tests below. Care to do the error check and if nothing else goes wrong, I'm ok with the `nil` though still slightly more comfy with the error (after all, even with parallelism, if your first request hasn't landed after five seconds, maybe it's time that the coordinator kills the txn itself).